core/trino-main/src/main/java/io/trino/server/security/oauth2/NimbusClient.java

```diff
@@ -243,6 +265,63 @@ private String hashNonce(String nonce)
         }
     }
 
+    private void initializeUsingDiscovery()
+    {
+        Failsafe.with(new RetryPolicy<>()
+                        .withMaxAttempts(-1)
+                        .withMaxDuration(discoveryTimeout)
+                        .withDelay(Duration.ofSeconds(1))
+                        .abortOn(IllegalStateException.class)
+                        .onFailedAttempt(attempt -> LOG.debug("OpenID Connect Metadata read failed: %s", attempt.getLastFailure())))
+                .get(() ->
+                        httpClient.execute(new OIDCProviderConfigurationRequest(issuer), response -> {
+                            int statusCode = response.getStatusCode();
+                            // stop on any client error other than REQUEST_TIMEOUT and TOO_MANY_REQUESTS
+                            if (400 <= statusCode && statusCode <= 499 && statusCode != REQUEST_TIMEOUT.code() && statusCode != TOO_MANY_REQUESTS.code()) {
+                                throw new RuntimeException("Client error: " + statusCode);
+                            }
+
+                            checkMetadataState(statusCode == HttpStatus.OK.code(), format("Expected response code to be %s, but was %d", HttpStatus.OK.code(), statusCode));
+                            JSONObject body = response.getContentAsJSONObject();
+
+                            OIDCProviderMetadata metadata = OIDCProviderMetadata.parse(body);
+                            checkMetadataState(issuer.equals(metadata.getIssuer()), "The value of the \"issuer\" claim in Metadata document different than the Issuer URL used for the Configuration Request.");
+
+                            authUrl = getRequiredField("authorization_endpoint", metadata.getAuthorizationEndpointURI(), AUTH_URL, authUrl);
```

Comment:
I don't like the way you overwrite the user input. I think user explicit configuration should overwrite the one from autoconfiguration.